includes/functions/llms.functions.access.php

```diff
@@ -31,149 +63,196 @@ function llms_page_restricted( $post_id, $user_id = null ) {
 		'restriction_id' => 0,
 	);
 
-	if ( ! $user_id ) {
-		$user_id = get_current_user_id();
-	}
-
-	$student = false;
-	if ( $user_id ) {
-		$student = new LLMS_Student( $user_id );
-	}
+	$student = $user_id ? llms_get_student( $user_id ) : false;
 
+	$post      = llms_get_post( $post_id );
 	$post_type = get_post_type( $post_id );
 
-	/**
-	 * Do checks to determine if the content should be restricted.
-	 */
 	$sitewide_membership_id = llms_is_post_restricted_by_sitewide_membership( $post_id, $user_id );
-	$membership_id          = llms_is_post_restricted_by_membership( $post_id, $user_id );
-
-	if ( is_home() && $sitewide_membership_id ) {
-		$restriction_id = $sitewide_membership_id;
-		$reason         = 'sitewide_membership';
-		// if it's a search page and the site isn't restricted to a membership bypass restrictions.
-	} elseif ( ( is_search() ) && ! get_option( 'lifterlms_membership_required', '' ) ) {
-		return apply_filters( 'llms_page_restricted', $results, $post_id );
-	} elseif ( is_singular() && $sitewide_membership_id ) {
```

Comment:
from here on this PR introduces a regression (bug) by removing the checks on `is_singular()`.
Because on `template_include` we run this function, see: https://github.com/gocodebox/lifterlms/blob/4.3.0/includes/class.llms.template.loader.php#L101

and you know what `get_the_ID()` returns on archives at `template_loader` => the first post ID of the list.
You can easily reproduce this bug:
1) Create a post A and assign it the newly created category Cat_1
2) Create a post B and assign it the category Cat_1, then limit this post to a membership.
3) Visit category Cat_1 archive page
You'll see no list and the restriction message. This is not what we want (although, we probably might want to filter the content/excerpt of the restricted post in archives/search as well, but out of the scope of this PR).

What can we do?
Probably we shouldn't check against `get_the_ID()` in the template loader class, but against `get_queried_object_id()`, what do you think?
We should stress out these edge cases.

Reply:
this introduces an interesting (related?) issue... We're passing in a $post_id but we're actually checking a "page's" (as in a generated screen's) not a post's restrictions.

In this scenario the queried object ID would be a category ID, right?

If we pass that in we should be adjusting the signature to accept an object id instead of a post id. And then, we need another argument to pass in the object type because we may have a category that shares an id with a post.

This would make the method more robust but it would also constitute some backwards incompatibilities, I think.

Maybe instead of doing this we can leave it as it is and add taxonomy archive checks before the "singular" post checks at the end...

Before this change what happens in your example? (I know I can load it up but I haven't yet)

Reply:
> this introduces an interesting (related?) issue... We're passing in a $post_id but we're actually checking a "page's" (as in a generated screen's) not a post's restrictions.

yeah 

> In this scenario the queried object ID would be a category ID, right?

right

> If we pass that in we should be adjusting the signature to accept an object id instead of a post id. And then, we need another argument to pass in the object type because we may have a category that shares an id with a post.

yeah, why not?

> This would make the method more robust but it would also constitute some backwards incompatibilities, I think.

Not sure it will, because it basically doesn't change the behavior... Before this PR categories were not affected by a restriction if not the sitewide one, or similar, for all the others there was the check on `is_singular()` that excluded them.

> Maybe instead of doing this we can leave it as it is and add taxonomy archive checks before the "singular" post checks at the end...
> 
well yeah we need something like that. You know that the nature of the current context is always tricky in wp, because of plugin's query alterations... it might really break our _compatibility balls_ :D 
You know if `! is_singular()` we don't need to know the `id` of the current object. If we pass a falsy post_id (which is not optional), we should only run those checks on the sitewide membership restriction, right?

> Before this change what happens in your example? (I know I can load it up but I haven't yet)

Well the category displayed all the posts...

Reply:
Now that i'm thinking about it catalogs are *not restricted* by design when sitewide membership is enabled.

So the real issue (I realize this now) is that we are restricting something incorrectly. `is_singular()` ensures that we don't run these checks against these "screens"

So I think we should change the behavior so that we do not run the formerly `is_singular()` checks when viewing archives of any kind.

Reply:
Oh shi... you're right about the catalogs... Yeah but why?
Well anyways, at this point, if we don't want catalogs to be restricted we just can stop calling `llms_page_restricted()` in  https://github.com/gocodebox/lifterlms/blob/4.3.0/includes/class.llms.template.loader.php#L101
right?

Reply:
> Yeah but why?

Since courses and memberships themselves bypass sitewide membership restrictions, so should the catalogs. The idea is that this is how a user determines what is available to them if they join the membership that provides them with access to the content.

If memberships cannot be found/discovered then it's impossible to join the closed site at all.

There have been a few requests to lock down *everything* including these areas of the site but not very many. I would chalk this up to the fact that this is a pretty misunderstood feature of LifterLMS in general (and one that I wish we never added as it's a niche feature that generally confuses users instead of empowers them to solve a problem).

In any event, whether or not this is the *correct* behavior is a different conversation. It is, however, the current behavior and I'd like to maintain that.

Reply:
I'm sorry, I thought that you were using "catalogs" in a more abstract way to identify any kind of archives, because we don't apply restrictions to any archives.

So yes I understood why our memberships and courses catalogs where not restricted, and my "but why?" question was more about other archives. And wanted to understand this to see whether doing what I said above made any sense:

> Well anyways, at this point, if we don't want catalogs to be restricted we just can stop calling llms_page_restricted() in https://github.com/gocodebox/lifterlms/blob/4.3.0/includes/class.llms.template.loader.php#L101
> right?

And for catalogs I wrongly meant all kind of archives.

Reply:
After reviewing this closer to try to fix the regression I've found a ton of inconsistencies in the rule(s) here with regards to sitewide memberships...

I think we can take this as an opportunity to possibly right some broken behavior and possibly make a new option alongside the sitewide membership restriction option to allow end users to customize this behavior.

My thinking is as follows, when a sitewide membership is enabled:

+ All site content (excluding the exceptions listed [here](https://github.com/gocodebox/lifterlms/blob/b54afbd510c7b1c6ad0465671d569e9044e42221/includes/functions/llms.functions.access.php#L630-L638) should be restricted. This includes the site's homepage, blog page, any post type archives, etc... Courses are restricted. Additionally, the membership catalog page should also be restricted. The exception pages exist to facilitate user self-onboarding onto the site. Users can view the (restricted) view of the membership and pages that support enrollment into that membership.
+ Add a textarea field where users can enter a list of URLs that they wish to add to this bypass list. Users can add one url per line. These URLs will be automatically excluded from restrictions.

The end result will be what I believe is supposed to be the current behavior but is not working properly (both without this PR and with it).

Thoughts?

Reply:
I'm with you.